File: engine/core/src/elements/sram.rs
//! Static RAM (SRAMElm.java, dump 413) and ROM (ROMElm.java, dump 436), one
//! shared memory model with a `has_we` switch. The address and data bits ride
//! on the west and east sides, WE and OE are active-low, and each data pin is
//! a bidirectional terminal: an internal voltage source holds the stored bit
//! and a per-step resistor couples it to the pin (1 ohm while output enabled,
//! else a 1e8 pulldown), so the pin can also be driven from outside during a
//! write (SRAMElm.java:283-331).
//!
//! The file line carries the `addressBits` and `dataBits` tokens after the
//! optional high voltage, then the stored contents as runs of consecutive
//! addresses: `addr val val ... -1 addr val ... -1 ... -2` (SRAMElm.java:
//! 55-70). Upstream's text `dump()` drops both the sizes and the contents,
//! so this port's writer restores them (the same quirk fix as the
//! thermistor's position token); the reader here matches the constructor.

use std::collections::HashMap;

use crate::element::{Base, Element, SimCtx};
use crate::spec::ElementSpec;
use crate::stamp::{Stamper, GROUND};

/// File flag saying the map reloads from its initial contents on reset
/// (SRAMElm.java:36).
const FLAG_RELOAD_ON_RESET: i64 = 2;

/// The write-time pulldown on each data pin, `stampResistor(data, ground, 1e8)`
/// (SRAMElm.java:315).
const WRITE_PULLDOWN_R: f64 = 1e8;
/// The output coupling resistance from the internal source to the data pin
/// (SRAMElm.java:313).
const OUTPUT_R: f64 = 1.0;

/// Reads the flat `addr{i}`/`val{i}` pair stream the frontend dumps into
/// params out of the file's run groups. Last write wins per address, matching
/// upstream's `map.put` over runs that overlap.
fn load_contents(spec: &ElementSpec) -> HashMap<usize, u32> {
    let mut map = HashMap::new();
    let mut i = 0usize;
    while let Some(&a) = spec.params.get(&format!("addr{i}")) {
        let v = spec.params.get(&format!("val{i}")).copied().unwrap_or(0.0);
        map.insert(a as usize, v as u32);
        i += 1;
    }
    map
}

/// SRAM (with a WE pin) and ROM (without one). The ROM is SRAM with the write
/// path removed: no WE pin, no `step_finished` write-back (ROMElm.java:90).
pub struct Sram {
    base: Base,
    high_voltage: f64,
    address_bits: usize,
    data_bits: usize,
    /// True under upstream's BIT_ORDER_BUS (ChipElm.java:37): the address
    /// bank and the data bank each share one coordinate, told apart by
    /// per-post bit tags.
    bus: bool,
    has_we: bool,
    map: HashMap<usize, u32>,
    /// The contents to restore under FLAG_RELOAD_ON_RESET, captured at build
    /// (SRAMElm.java:72-73).
    initial_map: Option<HashMap<usize, u32>>,
    /// The address decoded on the last step, kept for the write-back
    /// (`address`, SRAMElm.java:281).
    address: usize,
}

impl Sram {
    pub fn new(spec: &ElementSpec, has_we: bool) -> Self {
        // The edit dialog limits both widths to 2..16 (SRAMElm.java:228-241);
        // the clamp keeps a corrupt token from allocating unbounded posts.
        let address_bits = (spec.param("addressBits", 4.0) as usize).clamp(2, 16);
        let data_bits = (spec.param("dataBits", 4.0) as usize).clamp(2, 16);
        let map = load_contents(spec);
        let initial_map = if spec.flag(FLAG_RELOAD_ON_RESET) {
            Some(map.clone())
        } else {
            None
        };
        Self {
            base: Base::with_posts(Self::post_total(has_we, address_bits, data_bits)),
            high_voltage: spec.param("highVoltage", 5.0),
            address_bits,
            data_bits,
            bus: spec.flag(crate::elements::chip::FLAG_BIT_ORDER_BUS),
            has_we,
            map,
            initial_map,
            address: 0,
        }
    }

    /// Post count: WE + OE (or just OE for a ROM) plus the two pin banks
    /// (getPostCount, SRAMElm.java:124-126, ROMElm.java:60-62).
    fn post_total(has_we: bool, address_bits: usize, data_bits: usize) -> usize {
        let controls = if has_we { 2 } else { 1 };
        controls + address_bits + data_bits
    }

    fn oe_post(&self) -> usize {
        if self.has_we {
            1
        } else {
            0
        }
    }

    /// First address pin's post index (`addressNodes`, SRAMElm.java:117,
    /// ROMElm.java:51).
    fn address_base(&self) -> usize {
        if self.has_we {
            2
        } else {
            1
        }
    }

    /// First data pin's post index (`dataNodes`, SRAMElm.java:118).
    fn data_base(&self) -> usize {
        self.address_base() + self.address_bits
    }

    /// First internal node's local index, right after the posts
    /// (`internalNodes`, SRAMElm.java:119).
    fn internal_base(&self) -> usize {
        self.post_count()
    }

    fn threshold(&self) -> f64 {
        self.high_voltage * 0.5
    }

    /// Decodes the address pins MSB first (SRAMElm.java:299-302).
    fn read_address(&self) -> usize {
        let mut address = 0usize;
        for i in 0..self.address_bits {
            if self.base.volts[self.address_base() + i] > self.threshold() {
                address |= 1 << (self.address_bits - 1 - i);
            }
        }
        address
    }
}

impl Element for Sram {
    fn kind(&self) -> &'static str {
        if self.has_we {
            "sram"
        } else {
            "rom"
        }
    }
    fn base(&self) -> &Base {
        &self.base
    }
    fn base_mut(&mut self) -> &mut Base {
        &mut self.base
    }
    fn post_count(&self) -> usize {
        Self::post_total(self.has_we, self.address_bits, self.data_bits)
    }
    /// Bus mode: each bank collapses onto one coordinate, and makeBitPins
    /// runs the banks reversed, so bank pin p carries bit `n-1-p` of its
    /// bank, which is exactly what read_address (:132) and do_step (:210)
    /// decode (SRAMElm.java:120-121).
    fn post_bus_z(&self, post: usize) -> usize {
        if !self.bus {
            return 0;
        }
        let a0 = self.address_base();
        let d0 = self.data_base();
        if post >= a0 && post < d0 {
            self.address_bits - 1 - (post - a0)
        } else if post >= d0 && post < d0 + self.data_bits {
            self.data_bits - 1 - (post - d0)
        } else {
            0
        }
    }
    /// One node per data bit holds the source that drives the stored value
    /// (`getInternalNodeCount`, SRAMElm.java:268).
    fn internal_node_count(&self) -> usize {
        self.data_bits
    }
    /// One voltage source per data bit, from ground to the internal node
    /// (SRAMElm.java:267).
    fn voltage_source_count(&self) -> usize {
        self.data_bits
    }
    fn voltage_source_nodes(&self, k: usize) -> (usize, usize) {
        // setVoltageSource (SRAMElm.java:269-272): the source spans ground to
        // the internal node, never the data pin itself.
        (GROUND, self.base.nodes[self.internal_base() + k])
    }
    /// No posts couple in the matrix: the data pins reach the internal nodes
    /// only through the per-step resistor, which the `matrix_connects`
    /// override puts in one closure (ChipElm.java:467).
    fn connects(&self, _a: usize, _b: usize) -> bool {
        false
    }
    /// Each internal node couples to its data pin through the per-step output
    /// resistor (getMatrixConnection, SRAMElm.java:273-279).
    fn matrix_connects(&self, a: usize, b: usize) -> bool {
        (0..self.data_bits).any(|i| {
            let internal = self.internal_base() + i;
            let data = self.data_base() + i;
            (a == internal && b == data) || (a == data && b == internal)
        })
    }
    /// The resistor changes with the pin levels every step, so the element is
    /// nonlinear (nonLinear, SRAMElm.java:103).
    fn nonlinear(&self) -> bool {
        true
    }
    fn stamp(&mut self, _ctx: &SimCtx, s: &mut Stamper) {
        // Topology once with a zero value; `do_step` supplies the stored bits
        // and the coupling resistors (SRAMElm.java:283-291).
        for k in 0..self.data_bits {
            s.voltage_source(
                GROUND,
                self.base.nodes[self.internal_base() + k],
                self.base.vs_base + k,
                0.0,
            );
        }
    }
    fn do_step(&mut self, _ctx: &SimCtx, s: &mut Stamper) {
        let write_enabled = self.has_we && self.base.volts[0] < self.threshold();
        let output_enabled = !write_enabled && self.base.volts[self.oe_post()] < self.threshold();
        let address = self.read_address();
        let data = self.map.get(&address).copied().unwrap_or(0);
        for i in 0..self.data_bits {
            let bit = (data >> (self.data_bits - 1 - i)) & 1 == 1;
            s.voltage_source_value(
                self.base.vs_base + i,
                if bit { self.high_voltage } else { 0.0 },
            );
            // Output enabled: couple the internal source to the data pin so it
            // can drive the external net. Otherwise pull the pin down so an
            // external write driver can hold it (SRAMElm.java:306-316).
            if output_enabled {
                s.resistor(
                    self.base.nodes[self.internal_base() + i],
                    self.base.nodes[self.data_base() + i],
                    OUTPUT_R,
                );
            } else {
                s.resistor(
                    self.base.nodes[self.data_base() + i],
                    GROUND,
                    WRITE_PULLDOWN_R,
                );
            }
        }
        self.address = address;
    }
    fn step_finished(&mut self, _ctx: &SimCtx) {
        // A ROM has no write path (ROMElm.java:90). While WE is low the SRAM
        // samples the data pins and stores them (SRAMElm.java:319-331).
        if !self.has_we || self.base.volts[0] >= self.threshold() {
            return;
        }
        let mut data = 0u32;
        for i in 0..self.data_bits {
            if self.base.volts[self.data_base() + i] > self.threshold() {
                data |= 1 << (self.data_bits - 1 - i);
            }
        }
        self.map.insert(self.address, data);
    }
    fn calculate_current(&mut self, _ctx: &SimCtx) {
        self.base.current = 0.0;
    }
    fn current_into_node(&self, post: usize) -> f64 {
        let data_base = self.data_base();
        if post >= data_base && post < data_base + self.data_bits {
            let i = post - data_base;
            let write_enabled = self.has_we && self.base.volts[0] < self.threshold();
            let output_enabled =
                !write_enabled && self.base.volts[self.oe_post()] < self.threshold();
            if output_enabled {
                // The stored bit's source drives the data pin through the
                // 1 ohm coupling; the delivered current is what leaves the
                // element into the pin's node.
                (self.base.volts[self.internal_base() + i] - self.base.volts[post]) / OUTPUT_R
            } else {
                // The write-time pulldown draws a negligible leak out of the
                // externally driven pin.
                -self.base.volts[post] / WRITE_PULLDOWN_R
            }
        } else {
            0.0
        }
    }
    fn set_param(&mut self, name: &str, value: f64) -> bool {
        match name {
            "highVoltage" => self.high_voltage = value,
            // The widths change the post and node counts, which only a full
            // rebuild can reallocate.
            _ => return false,
        }
        true
    }
    fn reset(&mut self) {
        self.base.reset();
        self.address = 0;
        if let Some(initial) = &self.initial_map {
            // Restore the load-time contents under FLAG_RELOAD_ON_RESET
            // (SRAMElm.java:97-101).
            self.map = initial.clone();
        }
    }
}

File: web/src/model/registry/elements/sram.ts
/**
 * The static RAM (SRAMElm.java, dump 413): a chip whose address and data bits
 * ride on the west and east sides, with active-low WE and OE at the top. The
 * file line carries `addressBits dataBits` after the optional high voltage,
 * then the stored contents as runs of consecutive addresses: `addr val val
 * ... -1 addr val ... -1 ... -2` (SRAMElm.java:55-70). Upstream's text
 * `dump()` drops the sizes and the contents, so this port's writer restores
 * them (the same quirk fix as the thermistor's position token).
 *
 * The shared pin, parse and dump helpers serve the ROM too (dump 436), which
 * is the same memory chip without the WE pin; `rom.ts` builds its def on
 * them.
 */

import {
  CHIP_BIT_ORDER_BUS,
  chipBitOrderFlags,
  chipBitOrderParam,
  chipBodyRect,
  chipCommonTokens,
  chipDumpFlags,
  chipPosts,
  drawChip,
  normalizeChipBits,
  type ChipPinDef,
} from './dFlipFlop';
import type { CircuitElement, DrawContext, ElementDef } from '../../types';

/** The contents dialog's hex display (SRAMElm.java:30). Display-only: the
 *  token stream is decimal either way, so the bit just round-trips. */
export const SRAM_HEX_DISPLAY = 4;
/** Restore the load-time contents on reset (SRAMElm.java:36). */
export const SRAM_RELOAD_ON_RESET = 2;

/** The bit widths, floored like the engine: truncated and clamped to the
 *  2..16 the edit dialog allows (SRAMElm.java:228-241, sram.rs:79). */
export function normalizeSramBits(value: number): number {
  return normalizeChipBits(value, 2, 16);
}

function sramAddressBits(e: CircuitElement): number {
  return normalizeSramBits(e.params.addressBits ?? 4);
}

function sramDataBits(e: CircuitElement): number {
  return normalizeSramBits(e.params.dataBits ?? 4);
}

/** Bus bit order (upstream BIT_ORDER_BUS, XML attribute `bo="2"`): the
 *  address bank and the data bank each collapse onto one row of their side. */
export function memoryBus(e: CircuitElement): boolean {
  return e.params.bitOrder === 2 || (e.flags & CHIP_BIT_ORDER_BUS) !== 0;
}

/** `sizeY = max(addrY, dataY) + 1` (SRAMElm.java:110), with both bank heights
 *  1 in bus mode. */
export function memorySizeY(e: CircuitElement): number {
  if (memoryBus(e)) return 2;
  return Math.max(sramAddressBits(e), sramDataBits(e)) + 1;
}

/** The pin table, from `setupPins` (SRAMElm.java:106-123, ROMElm.java:37-58):
 *  the active-low WE (SRAM only) and OE at the top, then makeBitPins runs of
 *  address pins down the west and data pins down the east, both MSB first.
 *  The data pins are `output` for the post count and the draw only: they are
 *  bidirectional electrically, a role the engine models, and nothing is saved
 *  to the file. */
export function memoryPins(e: CircuitElement, hasWe: boolean): ChipPinDef[] {
  const ab = sramAddressBits(e);
  const db = sramDataBits(e);
  const bus = memoryBus(e);
  const sizeY = memorySizeY(e);
  const pins: ChipPinDef[] = [];
  if (hasWe) {
    pins.push({ side: 'W', pos: 0, text: 'WE', lineOver: true });
    pins.push({ side: 'E', pos: 0, text: 'OE', lineOver: true });
  } else {
    // The ROM's OE is the single west control pin, at the top-left
    // (ROMElm.java:47-49).
    pins.push({ side: 'W', pos: 0, text: 'OE', lineOver: true });
  }
  // makeBitPins reversed, MSB first: the list runs A_{ab-1}..A0 down the west
  // and D_{db-1}..D0 down the east, the MSB on the top row (SRAMElm.java:
  // 120-121). In bus mode each bank collapses onto row sizeY - bankY (both
  // bank heights are 1), every pin carrying its logical bit as its tag.
  const addrY = bus ? 1 : ab;
  const dataY = bus ? 1 : db;
  for (let j = 0; j < ab; j++) {
    pins.push({
      side: 'W',
      pos: sizeY - addrY + (bus ? 0 : j),
      text: bus ? 'A' : `A${ab - 1 - j}`,
      ...(bus ? { busWidth: ab, busZ: ab - 1 - j } : {}),
    });
  }
  for (let j = 0; j < db; j++) {
    pins.push({
      side: 'E',
      pos: sizeY - dataY + (bus ? 0 : j),
      text: bus ? 'D' : `D${db - 1 - j}`,
      output: true,
      ...(bus ? { busWidth: db, busZ: db - 1 - j } : {}),
    });
  }
  return pins;
}

/** Reads the contents runs starting at `t[start]` into the flat `addr{i}` /
 *  `val{i}` param pairs, in stream order. Grouping consecutive addresses on
 *  dump recovers the runs byte-exactly, and the engine's last-wins map insert
 *  over a pair list matches `map.put` over overlapping runs (SRAMElm.java:
 *  58-69). */
function readContentsRuns(t: string[], e: CircuitElement, start: number): void {
  let k = 0;
  let i = start;
  while (i < t.length) {
    const addr = Number(t[i]);
    if (!Number.isFinite(addr) || addr < 0) break;
    i++;
    if (i >= t.length) break;
    e.params[`addr${k}`] = addr;
    e.params[`val${k}`] = Number(t[i]);
    i++;
    k++;
    let next = addr;
    while (i < t.length) {
      const v = Number(t[i]);
      if (!Number.isFinite(v) || v < 0) {
        // Consume the run's -1 terminator so the next run-start read lands on
        // the following token (SRAMElm.java:66-69).
        i++;
        break;
      }
      next += 1;
      e.params[`addr${k}`] = next;
      e.params[`val${k}`] = v;
      i++;
      k++;
    }
  }
}

/** Reads the common chip tokens, the two width tokens and the contents runs.
 *  The layout is the same with or without the WE pin (ROMElm.java:28-31). */
export function memoryParse(t: string[], e: CircuitElement): void {
  const i = chipCommonTokens(t, e, false);
  const ab = Number(t[i]);
  if (t[i] !== undefined && Number.isFinite(ab)) e.params.addressBits = normalizeSramBits(ab);
  const db = Number(t[i + 1]);
  if (t[i + 1] !== undefined && Number.isFinite(db)) e.params.dataBits = normalizeSramBits(db);
  readContentsRuns(t, e, i + 2);
  chipBitOrderParam(e);
}

/** The flat `addr{i}` / `val{i}` params as an ordered pair list. The order is
 *  the stream order the file used, which is also the order `memoryDump`
 *  regroups runs from and the contents editor presents. */
export function memoryPairs(e: CircuitElement): [number, number][] {
  const pairs: [number, number][] = [];
  let k = 0;
  while (e.params[`addr${k}`] !== undefined) {
    pairs.push([e.params[`addr${k}`], e.params[`val${k}`] ?? 0]);
    k++;
  }
  return pairs;
}

/** Writes the common chip tokens, the two width tokens and the contents
 *  runs. Upstream's own `dump()` writes none of the memory's tokens, so this
 *  writer restores all of them: a save must not shrink the part to a blank
 *  4x4. */
export function memoryDump(e: CircuitElement): (string | number)[] {
  const out: (string | number)[] = [];
  const hv = e.params.highVoltage;
  if (hv !== undefined && hv !== 5) out.push(hv);
  out.push(e.params.addressBits ?? 4);
  out.push(e.params.dataBits ?? 4);
  const pairs = memoryPairs(e);
  // Regroup the flat pairs into runs of consecutive addresses.
  let i = 0;
  while (i < pairs.length) {
    out.push(pairs[i][0]);
    let expect = pairs[i][0];
    while (i < pairs.length && pairs[i][0] === expect) {
      out.push(pairs[i][1]);
      expect += 1;
      i += 1;
    }
    out.push(-1);
  }
  if (pairs.length > 0) out.push(-2);
  return out;
}

function drawSram(g: DrawContext, e: CircuitElement): void {
  drawChip(g, e, 2, memorySizeY(e), memoryPins(e, true));
}

export const SRAM_DEF: ElementDef = {
  kind: 'sram',
  label: 'static RAM',
  category: 'Logic',
  dumpCode: '413',
  postCount: 10,  // WE + OE + 2*4 bits at the default
  posts: (e) => chipPosts(e, 2, memorySizeY(e), memoryPins(e, true)),
  bodyRect: (e) => chipBodyRect(e, 2, memorySizeY(e)),
  noDiagonal: true,  // ChipElm.java:44
  defaultLength: 3,  // the chip spans (sizeX + 1) * 32
  defaults: { addressBits: 4, dataBits: 4, highVoltage: 5 },
  parse: (t, e) => memoryParse(t, e),
  dump: memoryDump,
  dumpFlags: (e) => chipBitOrderFlags(e, chipDumpFlags(e)),
  fields: [
    { name: 'addressBits', label: '# of Address Bits', min: 2, max: 16, integer: true },
    { name: 'dataBits', label: '# of Data Bits', min: 2, max: 16, integer: true },
    { name: 'contents', label: 'Contents', type: 'contents' },
    { name: 'highVoltage', label: 'High logic voltage', unit: 'V' },
    { name: 'hexDisplay', label: 'Hex Display', type: 'bool', flag: SRAM_HEX_DISPLAY },
    // The Load Contents From File row (SRAMElm.java:154, SRAMLoadFile
    // .java:31-48). No value rides the field: the picked bytes are encoded
    // into the contents run and committed through setMemoryContents, so
    // nothing here reaches params or the file line. The ROM has no such row.
    { name: 'loadFile', label: 'Load Contents From File', type: 'file', fileLoad: 'binary' },
    {
      name: 'reloadOnReset',
      label: 'Restore Contents on Reset',
      type: 'bool',
      flag: SRAM_RELOAD_ON_RESET,
    },
  ],
  draw: drawSram,
};
